### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [1]:
# Dependencies and Setup
import pandas as pd

# File to Load (Remember to Change These)
school_data_to_load = "Resources/schools_complete.csv"
student_data_to_load = "Resources/students_complete.csv"

# Read School and Student Data File and store into Pandas DataFrames
school_data = pd.read_csv(school_data_to_load)
student_data = pd.read_csv(student_data_to_load)

# Combine the data into a single dataset.  
school_data_complete = pd.merge(student_data, school_data, how="left", on=["school_name", "school_name"])

## District Summary

* Calculate the total number of schools

* Calculate the total number of students

* Calculate the total budget

* Calculate the average math score 

* Calculate the average reading score

* Calculate the percentage of students with a passing math score (70 or greater)

* Calculate the percentage of students with a passing reading score (70 or greater)

* Calculate the percentage of students who passed math **and** reading (% Overall Passing)

* Create a dataframe to hold the above results

* Optional: give the displayed data cleaner formatting

In [2]:
# Calculate the total number of schools
total_schools = school_data['School ID'].count()
total_schools

15

In [3]:
# Calculate the total number of students
total_students = student_data['Student ID'].count()
total_students

39170

In [4]:
# Calculate the total budget
total_budget = school_data['budget'].sum()
total_budget

24649428

In [5]:
# Calculate the average math score
ave_math = student_data['math_score'].mean()
ave_math

78.98537145774827

In [6]:
# Calculate the average reading score
ave_reading = student_data['reading_score'].mean()
ave_reading

81.87784018381414

In [7]:
# Calculate the percentage of students with a passing math score (70 or greater)
pass_math = student_data.loc[(student_data["math_score"]>=70)]
percent_pass_math = (pass_math["Student ID"].count() / total_students) * 100
percent_pass_math

74.9808526933878

In [8]:
# Calculate the percentage of students with a passing reading score (70 or greater)
pass_reading = student_data.loc[(student_data["reading_score"]>=70)]
percent_pass_reading = (pass_reading["Student ID"].count() / total_students) * 100
percent_pass_reading

85.80546336482001

In [9]:
# Calculate the percentage of students who passed math and reading (% Overall Passing)
pass_math_reading = student_data.loc[(student_data["math_score"]>=70) & (student_data["reading_score"]>=70)]
percent_pass_math_reading = (pass_math_reading["Student ID"].count() / total_students) * 100
percent_pass_math_reading

65.17232575950983

In [10]:
# Create a dataframe to hold the above results
district_summary_df = pd.DataFrame(
        {"Total Schools": [total_schools],
         "Total Students": [total_students],
         "Total Budget":[total_budget],
         "Average Math Score": [ave_math],
         "Average Reading Score": [ave_reading],
         "% Passing Math": [percent_pass_math],
         "% Passing Reading": [percent_pass_reading],
         "% Overall Passing": [percent_pass_math_reading]      
        })
district_summary_df

Total Schools  Total Students  Total Budget  Average Math Score  \
0             15           39170      24649428           78.985371   

   Average Reading Score  % Passing Math  % Passing Reading  % Overall Passing  
0               81.87784       74.980853          85.805463          65.172326

## School Summary

* Create an overview table that summarizes key metrics about each school, including:
  * School Name
  * School Type
  * Total Students
  * Total School Budget
  * Per Student Budget
  * Average Math Score
  * Average Reading Score
  * % Passing Math
  * % Passing Reading
  * % Overall Passing (The percentage of students that passed math **and** reading.)
  
* Create a dataframe to hold the above results

In [11]:
# Grouping the DataFrame by "school_name"
school_groups = school_data_complete.groupby("school_name")

In [12]:
# Total students
students_per_school = school_groups["Student ID"].count()

In [13]:
# Average Math Score
school_ave_math = school_groups["math_score"].mean()

In [14]:
# Average Reading Score
school_ave_reading = school_groups["reading_score"].mean()

In [15]:
# Remove students that didn't pass math and assign to a new dataframe
pass_math_df = school_data_complete.loc[(school_data_complete["math_score"]>=70)]

In [16]:
# Group by school to find the number of students passing math per school
pass_math_group = pass_math_df.groupby("school_name")
student_pass_math = pass_math_group["Student ID"].count()

In [17]:
# Remove students that didn't pass reading and assign to a new dataframe
pass_reading_df = school_data_complete.loc[(school_data_complete["reading_score"]>=70)]

In [18]:
# Group by school to find the number of students passing reading per school
pass_reading_group = pass_reading_df.groupby("school_name")
student_pass_reading = pass_reading_group["Student ID"].count()

In [19]:
# Remove students that didn't pass math AND reading and assign to a new dataframe
pass_both_df = school_data_complete.loc[(school_data_complete["math_score"]>=70) & (school_data_complete["reading_score"]>=70)]

In [20]:
# Group by school to find the number of students passing math AND reading per school
pass_both_group = pass_both_df.groupby("school_name")
student_pass_both = pass_both_group["Student ID"].count()

In [21]:
# Create a dataframe to hold the above results
school_df = pd.DataFrame({
                        "Total Students": students_per_school, 
                        "Average Math Score": school_ave_math, 
                        "Average Reading Score": school_ave_reading,
                        "Total Students Pass Math": student_pass_math,
                        "Total Students Pass Reading": student_pass_reading, 
                        "Total Students Pass Math & Reading": student_pass_both 
                         })

In [22]:
# Create a new column that calculates the percent students passing math
school_df["% Passing Math"] = (school_df["Total Students Pass Math"] /
                             school_df["Total Students"])*100

In [23]:
# Create a new column that calculates the percent students passing reading
school_df["% Passing Reading"] = (school_df["Total Students Pass Reading"] /
                             school_df["Total Students"])*100

In [24]:
# Create a new column that calculates the percent students passing math AND reading
school_df["% Overall Passing"] = (school_df["Total Students Pass Math & Reading"] /
                             school_df["Total Students"])*100

In [25]:
# Merge the new DataFrame to the school_data DataFrame provided to get school type and budget 
school_summary_df = pd.merge(school_data, school_df, on="school_name", how="left")

In [26]:
# Create a new column that calculates the per student budget
school_summary_df["Per Student Budget"] = (school_summary_df["budget"] / school_summary_df["Total Students"])

In [27]:
# Rename columns
school_summary_df = school_summary_df.rename(columns={"school_name": "School Name",
                                                         "type": "School Type",
                                                         "budget": "Total School Budget"})

In [28]:
# Organise columns
school_summary_df = school_summary_df[["School Name", 
                                          "School Type", 
                                          "Total Students", 
                                          "Total School Budget",
                                          "Per Student Budget",
                                          "Average Math Score",
                                          "Average Reading Score",
                                          "% Passing Math",
                                          "% Passing Reading",
                                          "% Overall Passing"
                                         ]]

In [29]:
# Reset index to School Name
school_summary_df = school_summary_df.set_index('School Name')
school_summary_df

School Type  Total Students  Total School Budget  \
School Name                                                              
Huang High School        District            2917              1910635   
Figueroa High School     District            2949              1884411   
Shelton High School       Charter            1761              1056600   
Hernandez High School    District            4635              3022020   
Griffin High School       Charter            1468               917500   
Wilson High School        Charter            2283              1319574   
Cabrera High School       Charter            1858              1081356   
Bailey High School       District            4976              3124928   
Holden High School        Charter             427               248087   
Pena High School          Charter             962               585858   
Wright High School        Charter            1800              1049400   
Rodriguez High School    District            3999              2547363   
Johnson High School      District            4761              3094650   
Ford High School         District            2739              1763916   
Thomas High School        Charter            1635              1043130   

                       Per Student Budget  Average Math Score  \
School Name                                                     
Huang High School                   655.0           76.629414   
Figueroa High School                639.0           76.711767   
Shelton High School                 600.0           83.359455   
Hernandez High School               652.0           77.289752   
Griffin High School                 625.0           83.351499   
Wilson High School                  578.0           83.274201   
Cabrera High School                 582.0           83.061895   
Bailey High School                  628.0           77.048432   
Holden High School                  581.0           83.803279   
Pena High School                    609.0           83.839917   
Wright High School                  583.0           83.682222   
Rodriguez High School               637.0           76.842711   
Johnson High School                 650.0           77.072464   
Ford High School                    644.0           77.102592   
Thomas High School                  638.0           83.418349   

                       Average Reading Score  % Passing Math  \
School Name                                                    
Huang High School                  81.182722       65.683922   
Figueroa High School               81.158020       65.988471   
Shelton High School                83.725724       93.867121   
Hernandez High School              80.934412       66.752967   
Griffin High School                83.816757       93.392371   
Wilson High School                 83.989488       93.867718   
Cabrera High School                83.975780       94.133477   
Bailey High School                 81.033963       66.680064   
Holden High School                 83.814988       92.505855   
Pena High School                   84.044699       94.594595   
Wright High School                 83.955000       93.333333   
Rodriguez High School              80.744686       66.366592   
Johnson High School                80.966394       66.057551   
Ford High School                   80.746258       68.309602   
Thomas High School                 83.848930       93.272171   

                       % Passing Reading  % Overall Passing  
School Name                                                  
Huang High School              81.316421          53.513884  
Figueroa High School           80.739234          53.204476  
Shelton High School            95.854628          89.892107  
Hernandez High School          80.862999          53.527508  
Griffin High School            97.138965          90.599455  
Wilson High School             96.539641          90.582567  
Cabrera High School            97.039828          91.334769  
Bailey Hig

## Top Performing Schools (By % Overall Passing)

* Sort and display the top five performing schools by % overall passing.

In [30]:
top_schools_df = school_summary_df.sort_values("% Overall Passing", ascending=False)
top_schools_df.head(5)

School Type  Total Students  Total School Budget  \
School Name                                                            
Cabrera High School     Charter            1858              1081356   
Thomas High School      Charter            1635              1043130   
Griffin High School     Charter            1468               917500   
Wilson High School      Charter            2283              1319574   
Pena High School        Charter             962               585858   

                     Per Student Budget  Average Math Score  \
School Name                                                   
Cabrera High School               582.0           83.061895   
Thomas High School                638.0           83.418349   
Griffin High School               625.0           83.351499   
Wilson High School                578.0           83.274201   
Pena High School                  609.0           83.839917   

                     Average Reading Score  % Passing Math  % Passing Reading  \
School Name                                                                     
Cabrera High School              83.975780       94.133477          97.039828   
Thomas High School               83.848930       93.272171          97.308869   
Griffin High School              83.816757       93.392371          97.138965   
Wilson High School               83.989488       93.867718          96.539641   
Pena High School                 84.044699       94.594595          95.945946   

                     % Overall Passing  
School Name                             
Cabrera High School          91.334769  
Thomas High School           90.948012  
Griffin High School          90.599455  
Wilson High School           90.582567  
Pena High School             90.540541

## Bottom Performing Schools (By % Overall Passing)

* Sort and display the five worst-performing schools by % overall passing.

In [31]:
worst_schools_df = school_summary_df.sort_values("% Overall Passing")
worst_schools_df.head(5)

School Type  Total Students  Total School Budget  \
School Name                                                              
Rodriguez High School    District            3999              2547363   
Figueroa High School     District            2949              1884411   
Huang High School        District            2917              1910635   
Hernandez High School    District            4635              3022020   
Johnson High School      District            4761              3094650   

                       Per Student Budget  Average Math Score  \
School Name                                                     
Rodriguez High School               637.0           76.842711   
Figueroa High School                639.0           76.711767   
Huang High School                   655.0           76.629414   
Hernandez High School               652.0           77.289752   
Johnson High School                 650.0           77.072464   

                       Average Reading Score  % Passing Math  \
School Name                                                    
Rodriguez High School              80.744686       66.366592   
Figueroa High School               81.158020       65.988471   
Huang High School                  81.182722       65.683922   
Hernandez High School              80.934412       66.752967   
Johnson High School                80.966394       66.057551   

                       % Passing Reading  % Overall Passing  
School Name                                                  
Rodriguez High School          80.220055          52.988247  
Figueroa High School           80.739234          53.204476  
Huang High School              81.316421          53.513884  
Hernandez High School          80.862999          53.527508  
Johnson High School            81.222432          53.539172

## Math Scores by Grade

* Create a table that lists the average Reading Score for students of each grade level (9th, 10th, 11th, 12th) at each school.

  * Create a pandas series for each grade. Hint: use a conditional statement.
  
  * Group each series by school
  
  * Combine the series into a dataframe
  
  * Optional: give the displayed data cleaner formatting

In [32]:
# Use .loc to filter each grade
ninth_grade = school_data_complete.loc[school_data_complete["grade"] == "9th"]
tenth_grade = school_data_complete.loc[school_data_complete["grade"] == "10th"]
eleventh_grade = school_data_complete.loc[school_data_complete["grade"] == "11th"]
twelfth_grade = school_data_complete.loc[school_data_complete["grade"] == "12th"]

In [33]:
# Group each DataFrame by school
ninth_grade_groups = ninth_grade.groupby("school_name")
tenth_grade_groups = tenth_grade.groupby("school_name")
eleventh_grade_groups = eleventh_grade.groupby("school_name")
twelfth_grade_groups = twelfth_grade.groupby("school_name")

In [34]:
# Calculate average math score by school for each grade level
ninth_grade_ave_math = ninth_grade_groups['math_score'].mean()
tenth_grade_ave_math = tenth_grade_groups['math_score'].mean()
eleventh_grade_ave_math = eleventh_grade_groups['math_score'].mean()
twelfth_grade_ave_math = twelfth_grade_groups['math_score'].mean()

In [35]:
# Combine the series into a dataframe
school_grade_ave_math_df = pd.DataFrame({"9th": ninth_grade_ave_math, 
                        "10th": tenth_grade_ave_math, 
                        "11th": eleventh_grade_ave_math,
                        "12th": twelfth_grade_ave_math
                         })
school_grade_ave_math_df

9th       10th       11th       12th
school_name                                                      
Bailey High School     77.083676  76.996772  77.515588  76.492218
Cabrera High School    83.094697  83.154506  82.765560  83.277487
Figueroa High School   76.403037  76.539974  76.884344  77.151369
Ford High School       77.361345  77.672316  76.918058  76.179963
Griffin High School    82.044010  84.229064  83.842105  83.356164
Hernandez High School  77.438495  77.337408  77.136029  77.186567
Holden High School     83.787402  83.429825  85.000000  82.855422
Huang High School      77.027251  75.908735  76.446602  77.225641
Johnson High School    77.187857  76.691117  77.491653  76.863248
Pena High School       83.625455  83.372000  84.328125  84.121547
Rodriguez High School  76.859966  76.612500  76.395626  77.690748
Shelton High School    83.420755  82.917411  83.383495  83.778976
Thomas High School     83.590022  83.087886  83.498795  83.497041
Wilson High School     83.085578  83.724422  83.195326  83.035794
Wright High School     83.264706  84.010288  83.836782  83.644986

## Reading Score by Grade 

* Perform the same operations as above for reading scores

In [38]:
# Calculate average reading score by school for each grade level
ninth_grade_ave_reading = ninth_grade_groups['reading_score'].mean()
tenth_grade_ave_reading = tenth_grade_groups['reading_score'].mean()
eleventh_grade_ave_reading = eleventh_grade_groups['reading_score'].mean()
twelfth_grade_ave_reading = twelfth_grade_groups['reading_score'].mean()

In [40]:
# Combine the series into a dataframe
school_grade_ave_reading_df = pd.DataFrame({"9th": ninth_grade_ave_reading, 
                        "10th": tenth_grade_ave_reading, 
                        "11th": eleventh_grade_ave_reading,
                        "12th": twelfth_grade_ave_reading
                         })
school_grade_ave_reading_df

9th       10th       11th       12th
school_name                                                      
Bailey High School     81.303155  80.907183  80.945643  80.912451
Cabrera High School    83.676136  84.253219  83.788382  84.287958
Figueroa High School   81.198598  81.408912  80.640339  81.384863
Ford High School       80.632653  81.262712  80.403642  80.662338
Griffin High School    83.369193  83.706897  84.288089  84.013699
Hernandez High School  80.866860  80.660147  81.396140  80.857143
Holden High School     83.677165  83.324561  83.815534  84.698795
Huang High School      81.290284  81.512386  81.417476  80.305983
Johnson High School    81.260714  80.773431  80.616027  81.227564
Pena High School       83.807273  83.612000  84.335938  84.591160
Rodriguez High School  80.993127  80.629808  80.864811  80.376426
Shelton High School    84.122642  83.441964  84.373786  82.781671
Thomas High School     83.728850  84.254157  83.585542  83.831361
Wilson High School     83.939778  84.021452  83.764608  84.317673
Wright High School     83.833333  83.812757  84.156322  84.073171

## Scores by School Spending

* Create a table that breaks down school performances based on average Spending Ranges (Per Student). Use 4 reasonable bins to group school spending. Include in the table each of the following:
  * Average Math Score
  * Average Reading Score
  * % Passing Math
  * % Passing Reading
  * Overall Passing Rate (Average of the above two)

In [41]:
# Create 4 bins to hold school spending
bins = [0, 583, 629, 644, 675]

# Create the names for the 4 bins
bin_names = ["<$584", "$584-629", "$630-644", "$645-675"]

In [42]:
# Slice the data and place it into bins
# Place the data series into a new column inside of the DataFrame
school_summary_df["Spending Ranges (Per Student)"] = pd.cut(school_summary_df["Per Student Budget"], bins, labels=bin_names)

In [43]:
# Create a GroupBy object based upon "Spending Ranges (Per Student)"
school_spending_group = school_summary_df.groupby("Spending Ranges (Per Student)")

# Get the average of each column within the GroupBy object
school_spending_df = school_spending_group[["Average Math Score", "Average Reading Score", "% Passing Math", "% Passing Reading", "% Overall Passing"]].mean()

In [44]:
# Round average scores values to 2 decimals 
school_spending_df["Average Math Score"] = school_spending_df["Average Math Score"].map("{:.2f}".format)
school_spending_df["Average Reading Score"] = school_spending_df["Average Reading Score"].map("{:.2f}".format)

# Round percent students passing to no decimals
school_spending_df["% Passing Math"] = school_spending_df["% Passing Math"].map("{:.0f}".format)
school_spending_df["% Passing Reading"] = school_spending_df["% Passing Reading"].map("{:.0f}".format)
school_spending_df["% Overall Passing"] = school_spending_df["% Overall Passing"].map("{:.0f}".format)
school_spending_df

Average Math Score Average Reading Score  \
Spending Ranges (Per Student)                                            
<$584                                      83.46                 83.93   
$584-629                                   81.90                 83.16   
$630-644                                   78.52                 81.62   
$645-675                                   77.00                 81.03   

                              % Passing Math % Passing Reading  \
Spending Ranges (Per Student)                                    
<$584                                     93                97   
$584-629                                  87                93   
$630-644                                  73                84   
$645-675                                  66                81   

                              % Overall Passing  
Spending Ranges (Per Student)                    
<$584                                        90  
$584-629                                     81  
$630-644                                     63  
$645-675                                     54

## Scores by School Size

* Perform the same operations as above, based on school size.

In [45]:
# Create 4 bins to hold school spending
bins = [0, 999, 1999, 4999]

# Create the names for the 4 bins
bin_names = ["Small (<1000)", "Medium (1000-2000)", "Large (2000-5000)"]

In [46]:
# Slice the data and place it into bins
# Place the data series into a new column inside of the DataFrame
school_summary_df["School Size"] = pd.cut(school_summary_df["Total Students"], bins, labels=bin_names)

In [47]:
# Create a GroupBy object based upon "School Size"
school_size_group = school_summary_df.groupby("School Size")

# Get the average of each column within the GroupBy object
school_size_df = school_size_group[["Average Math Score", "Average Reading Score", "% Passing Math", "% Passing Reading", "% Overall Passing"]].mean()

In [48]:
# Round average scores values to 2 decimals 
school_size_df["Average Math Score"] = school_size_df["Average Math Score"].map("{:.2f}".format)
school_size_df["Average Reading Score"] = school_size_df["Average Reading Score"].map("{:.2f}".format)

# Round percent students passing to no decimals
school_size_df["% Passing Math"] = school_size_df["% Passing Math"].map("{:.0f}".format)
school_size_df["% Passing Reading"] = school_size_df["% Passing Reading"].map("{:.0f}".format)
school_size_df["% Overall Passing"] = school_size_df["% Overall Passing"].map("{:.0f}".format)
school_size_df

Average Math Score Average Reading Score % Passing Math  \
School Size                                                                  
Small (<1000)                   83.82                 83.93             94   
Medium (1000-2000)              83.37                 83.86             94   
Large (2000-5000)               77.75                 81.34             70   

                   % Passing Reading % Overall Passing  
School Size                                             
Small (<1000)                     96                90  
Medium (1000-2000)                97                91  
Large (2000-5000)                 83                58

## Scores by School Type

* Perform the same operations as above, based on school type

In [49]:
# Create a GroupBy object based on "School Type"
school_type_group = school_summary_df.groupby("School Type")

# Get the average of each column within the GroupBy object
school_type_df = school_type_group[["Average Math Score", "Average Reading Score", "% Passing Math", "% Passing Reading", "% Overall Passing"]].mean()

In [50]:
# Round average scores values to 2 decimals 
school_type_df["Average Math Score"] = school_type_df["Average Math Score"].map("{:.2f}".format)
school_type_df["Average Reading Score"] = school_type_df["Average Reading Score"].map("{:.2f}".format)

# Round percent students passing to no decimals
school_type_df["% Passing Math"] = school_type_df["% Passing Math"].map("{:.0f}".format)
school_type_df["% Passing Reading"] = school_type_df["% Passing Reading"].map("{:.0f}".format)
school_type_df["% Overall Passing"] = school_type_df["% Overall Passing"].map("{:.0f}".format)
school_type_df

Average Math Score Average Reading Score % Passing Math  \
School Type                                                           
Charter                  83.47                 83.90             94   
District                 76.96                 80.97             67   

            % Passing Reading % Overall Passing  
School Type                                      
Charter                    97                90  
District                   81                54

Description of at least two observable trends based on the data: 
Overall, 65% of students passed both math and reading. The percent of students passing math is lower (75%) compared to the percent passing reading (86%). 

Schools that spent the highest per student (645−675 dollars) had the lowest average percent of students overall passing (54%). While schools that spent the lowest per student (less than 584 dollars) had the highest percent of students overall passing (90%). Large school size (2000-5000 students) had the lowest average percent overall passing (58%). 

The top five performing schools by percent overall passing are charter schools. While the bottom five performing schools by % overall passing are district schools. Overall, district schools had a lower average percent of students pasing math (67%), reading (81%), and overall passing (54%) compared to charter schools.